activesupport/lib/active_support/abstract_class.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module AbstractClass
+    extend ActiveSupport::Concern
+
+    included do
+      if defined?(@root_class)
+        raise AbstractClassError, "#{self} is already part of #{root_class}'s hierarchy"
```

Comment:
Not sure if we should define a new error, or use something like `ActiveSupportError` 🤷‍♂️